Documentation/design-docs/finally-optimizations.md

```diff
@@ -175,6 +175,41 @@ G_M60484_IG06:
 Empty finally removal is unconditionally profitable: it should always
 reduce code size and improve code speed.
 
+Empty Try Removal
+---------------------
+
+If the try region of a try-finally is empty, and the jitted code will
+execute on a runtime that does not protect finally execution from
+thread abort, then the try-finally can be replaced with just the
+content of the finally.
+
+Empty trys with non-empty finallys often exist in code that must run
+under both thread-abort aware and non-thread-abort aware runtimes. In
+the former case the placement of cleanup code in thea finally ensures
+that the cleanup code will execute fully. But if thread abort is not
+possible, the extra protection offered by the finally is not needed.
+
+Empty try removal looks for try-finallys where the try region does
+nothing except invoke the finally. There are currently two different
+EH implementation models, so the try screening has two cases:
+
+* callfinally thunks (all but x86): the try must be a single empty
+basic block that always jumps to a callfinally that is the first
+half of a callalways pair;
+* non-callfinally thunks: the try must be a callalways pair where
+the first block is an empty callfinally
+
+The screening then verifies that the callfinally identified above is
+the only callfinally for the try. No other callfinallys are expected
+because this try cannot have multple leaves and its handler cannot be
+reached by nested exit paths.
+
+When the empty try is identified, the jit modifies the callaways pair
+to branch to the handler, modifies the handler's return points to
+branch directly to the continuation (the branch target of the second
+half of the callalways pair), updates various status flags on the
+blocks, and then removes the try-finally region.
+
```

Comment:
In the "callfinally thunks" case, you'll have `always->callfinally`, converted to `always->always`. Can you just get rid of the thunk entirely and redirect this thunk always?